# LiveCoding: Introduction to Python (solutions)

by [Luciano Gabbanelli](https://www.linkedin.com/in/luciano-gabbanelli-ph-d-75302218)

<img width=80 src="https://media.giphy.com/media/KAq5w47R9rmTuvWOWa/giphy.gif">

<img width=150 src="Images/Assembler.png">

## Zen of Python

In [2]:
!python --version

Python 3.9.12


In [3]:
! jupyter notebook --version

6.4.12


In [4]:
import this # "guiding principles" for writing computer programs

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


<img width=300 style="float: left;" src="https://media.giphy.com/media/3s0QuxoSX6DgdnGFoE/giphy.gif">

## Built-in objects

Python has some simple **built−in types**, e.g.
* **int**: integer $(\mathbb Z)$
* **float**: decimal numbers $($representing $\mathbb R)$
* **complex**: they can be thought as a 2-D plane $(xy-plane \rightarrow ReIm-plane,\,$ representing $\ \mathbb C)$
* **str**: strings, i.e. texts
* **bool**: booleans, represent the two truth values of logic and Boolean algebra

It has also some complex **built−in types** used to store collections of data, e.g. 
* **List**: store multiple items in a single variable. It is a collection which is ordered and changeable. Allows duplicate members.
* **Dict**: dictionaries, store data values in key:value pairs. It is a collection which is ordered (only from v. 3.7 onwards), changeable and do not allow duplicates.
* **Tuple**: it is a collection which is ordered and immutable. Allows duplicate members. Tuples are indexed.
* **Set**: it is a collection which is unordered, with unchangeable elements (items are, but you can remove items and add new ones), and unindexed. No duplicate members. Set intems cannot be referred to by index or key.

In [5]:
# Cell for discusing type. Enter some objects and check their type: 
4

4

In [6]:
type(4)

int

In [7]:
type(4.5)

float

In [8]:
type("4")

str

_In Python, if it is in quotation marks, then it is a string!_

In [9]:
type(Luciano)

NameError: name 'Luciano' is not defined

In [10]:
type("l.gabbanelli@assemblerschool.com")

str

In [11]:
# Why then is this well defined?
type(1j)

complex

***Booleans*** are a data type that takes either `True` or `False` values.

In [12]:
# Run this cell: 
False

False

In [13]:
# Enter a sentence that results in a boolean:
6 > 5

True

Shortcuts!  In command mode (blue), press
* C :   copy the cell ( ≠ ctrl + C )
* V :   paste the cell ( ≠ ctrl + V )

In [14]:
# Find out the `type()` of the previous statement. This time type `ty` and press tab key
type(4 > 5)

bool

In [15]:
# What do you expect this cell to return?
Verdad

NameError: name 'Verdad' is not defined

**Built−in types used to store collections of data:**

In [16]:
# Cell to discuss. Enter some objects: 
type([1,2,3])

list

In [17]:
type((1,'apple',3))

tuple

In [18]:
type({1,2,2,3})

set

In [19]:
{1,2,2,3}

{1, 2, 3}

In [20]:
planet_earth = {
    "name": "Earth",
    "surface_gravity": 9.8, # in m/s
    "is_there_life": True
}

In [21]:
planet_earth

{'name': 'Earth', 'surface_gravity': 9.8, 'is_there_life': True}

In [22]:
type(planet_earth)

dict

### Variables to store objects

Variables are ways of storing information. Objects can be stored in variables

> **Syntax:**   `name_of_variable = object`

**Conventions and style:**
- A variable name must start with a letter or the underscore character.
- A variable name can only contain alpha-numeric characters and underscores (A−z, 0−9, and _ ).
- A variable name cannot start with a number.
- Variable names are case-sensitive (age, Age and AGE are three different variables).
- It is good programming practice for variable and function names to be in lowercase, with words separated by underscores to improve readability.

**Task:** On Earth, the acceleration due to gravity is 9.8 m/s<sup>2</sup>, if the value of the acceleration on the Moon's surface is about 16.6% of the acceleration on the Earth's surface, what is its value? Round the result to two decimal places.

In [23]:
acc_earth = 9.8

$ \text{100%}\qquad \vcenter{\rule{2cm}{.5pt}}\qquad \text{acc_earth = 9.8 m/s}^2   $

$ \text{  16% }\qquad \vcenter{\rule{2cm}{.5pt}}\qquad \text{acc_moon = ?} $

Then,

$$\text{acc_moon} \quad=\quad \frac{16\text{%}\ *\ 9.8 \text{m/s}^2}{100\text{%}} $$

In [24]:
# Enter your code here:
acc_moon = acc_earth * 0.166
print(acc_moon)

1.6268000000000002


In [25]:
acc_moon # Other option is just to call for the variable, i.e. ac_moon (without print). The output is equivalent.

1.6268000000000002

In [26]:
# Tip: use the round() function. Returns the float number rounded to the given decimal places as input.
# By default, the number of decimals is 0, meaning that the function will return the nearest integer.
round(acc_moon, 2)

1.63

In [27]:
type(acc_moon)

float

What is the value of the variable `acc_moon` ?  1.6268000000000002 or 1.63 ?

> - By the way, to denote a word or phrase as `fancy_code`, enclose it in backticks ( \` ).
>
> - You can use also *Italics* (using \* or \_), __Bold__ (using two \*\* or __) or ***both*** (with three \*\*\*).
>
> - Use \> for blockquotes
    > - and \- (or \* or \+) for itemize.
        > - By convention, always use the same character for items.
        > - Indentation changes the item symbol.
        > - 4 spaces $\equiv$ 1 tab.
>
> - Inline $\LaTeX$ formulas makes your *Notebook* **awsome**: 

$$x'(t) = \lim_{\Delta t\to 0} \frac{x(t+\Delta t)-x(t)}{\Delta t} $$

## The `print()` function

This function prints the specified message.
The message can be any object, but it will be converted into a string before written to the screen.

> **Syntax:** `print(object(s), sep=separator)`
>
> **object**: as many as you like
>
> **sep** = 'separator': Optional. Specify how to separate the objects. Default is `' '` 


More on ***shortcuts***!

Press shift + tab inside the print() parenthesis:

* 1 time   &rarr;   documentation
* 2 times  &rarr;   extended documentation
* 4 times  &rarr;   documentation in a pager

In [28]:
print('Welcome', 'to', 'Assembler')

Welcome to Assembler


- Print a message indicating the acceleration at the earth's surface:

In [29]:
# Try \n and \t in the print function. What are they ment for?
# Type the code here:
print('Earth acceleration',acc_earth, sep='\n\t=\n\t\t', end = ' m/s\u00b2')

Earth acceleration
	=
		9.8 m/s²

- Define two variables with your name and age and print them on the screen:

In [30]:
# Type the code here:
name = 'Luciano'
age = 18
print('My name is ' + name + ' and I am ' + str(age) + ' years old.')

My name is Luciano and I am 18 years old.


### Old string formatting

Take a look at [Old-string-formatting](https://docs.python.org/3/library/stdtypes.html#old-string-formatting)

- Print the value of π rounded to two decimal places using the old string format.

**Tip:** In the programming world, a library is a collection of related modules (precompiled codes) that can be used later on in a program for some specific well-defined operations. Some useful libraries are, for example, math, NumPy, Pandas, etc..

You can find the π number in one of those libraries ;)

In [31]:
# Less common mathematical functions are defined inside libraries.
# Type the code here:
import math
print('The value of pi (%s) is approximately %.2f.' % ('\u03C0', math.pi)) # f stands for floating point decimal format, .2 the decimals.

The value of pi (π) is approximately 3.14.


### The `str.format()` method

Fancier output formatting: literal text may be accompanied by "replacement fields" delimited by braces (or curly brackets) { }. Each replacement field is replaced with the string value of the corresponding argument.

- Print the accelerations on the Earth and on the Moon using the `str.format()` method:

In [32]:
# Type the code here:
print('The acceleration of the {0} is {1} m/s\u00b2, while it is {2} m/s\u00b2 on the {satellite} '.format(
    'Earth', acc_earth, acc_moon, satellite='Moon'))

The acceleration of the Earth is 9.8 m/s², while it is 1.6268000000000002 m/s² on the Moon 


### Formatted string literals or, simply, f−strings

These format joined the party in Python 3.6. 
To use it, enter an 'f' (or 'F') before the opening quotation mark. Expressions between {variable} refer to variables. While other string literals always have a constant value, formatted strings are really expressions evaluated at run time.

In [33]:
f"{2 * 37}"

'74'

- Print the acceleration on Earth using f-strings:

In [34]:
# Type the code here:

print(f'Earth acceleration: {acc_earth} m/s\u00b2')

Earth acceleration: 9.8 m/s²


### Extras

A review on old-school string formating and f-strings can be found in [An Improved String Formatting Syntax (Guide)](https://realpython.com/python-f-strings/).

Besides, you can read all about f-strings it in [PEP 498](https://peps.python.org/pep-0498/).

## The `input()` function 

**`input()`** function

This function takes user input. By default, it returns the user input in form of a string.

> **Syntax**: `input(prompt)`
>
> **Parameter** is a prompt: a string representing a default message indicating the required input (optional)
>
> **Return**: a string object

**Task:** Request name and age as input and print them on screen.

In [1]:
# Type the code here:
name = input('Please, enter your name: ')
age = input('and now your age: ')

print("The name of the user is {} and his/her age is {}".format(name, age))

Please, enter your name: Luciano
and now your age: 18
The name of the user is Luciano and his/her age is 18


## Conversion between `float`, `int` and `str`

In [28]:
# Type the code here:
a = 11
type(a)

int

In [31]:
a = float(a)

In [32]:
type(a)

float

In [33]:
float("4r")

ValueError: could not convert string to float: '4r'

In [34]:
int(3.5)

3

In [35]:
int("24")

24

In [36]:
str(4*4.5)

'18.0'

**Task:** Enter two numbers as input and calculate the sum between them. Print the result on the screen.

*Watch out!* Remember how string addition works, is that really what you want?

In [11]:
# Type the code here:
num1 = int(input("Enter the first number: "))
num2 = int(input("Enter the second one: "))

addition = num1 + num2

print('The sum of the two given numbers is {}'.format(addition))

Enter the first number: 6
Enter the second one: 4
The sum of the two given numbers is 10


## Built−in functions and operators

### Basic arithmetic operators

| Arithmetic Operators | Description  |
| :---: | :---: |
| `+` | addition |
| `-` | subtraction |
| `*` | multiplication |
| `/` | division |
| `//` | floor division |
| `%` | modulus |
| `**` | power / exponentiation |

Let `a`, `b` and `c` be two variables (`a` and `b` integers, and `c` a floating point number), apply the last four operators on them, and interpret:

In [36]:
# Let us define 3 variables a, b, c.
# Type the code here:
a = 11
b = 2
c = -0.5

Operators:

In [37]:
# Divide them.
# Type the code here:
a / b

5.5

In [38]:
# Floor division. The double slash operator does not return the fractional part of the floating-point result.
# Type the code here:
a // b

5

In [39]:
# The modulus operator return the remainder of a division.
# Type the code here:
a % b

1

In [40]:
# Exponentiation.
# Type the code here:
a ** b

121

Some more examples?

In [41]:
# Absolute value.
# Type the code here:
abs(c)

0.5

In [42]:
# As we have imported math libraty before, we do not need to call it again.
math.log10(100)

2.0

Because

$$\log_{a}b=c \quad\Longleftrightarrow\quad a^c=b$$

***

More details can be found in the documentation. Get used to using it!!!

[**Python Math Library**](https://docs.python.org/3/library/math.html)

***


### Comparison (relational) operators

Compare numbers or strings and decide the relation among them.
These operators do not return a numeric value like arithmetic expressions do, but instead return 1, which represents **True**, or 0, which represents **False**.

| Relational Operators | Description  |
| :---: | :---: |
| `==` | equal to. If the values are equal, then True |
| `!=` | NOT equal to. If values of two operands are not equal, then condition becomes True.  |
| `< , >` | smaller than, greater than |
| `<= , >=` | smaller than or equal to, greater than or equal to |

In [43]:
# What do you expect from this command?
4 = 3

SyntaxError: cannot assign to literal (1872368816.py, line 1)

In [44]:
# equal to
# Type the code here:
a == 13

False

In [45]:
# != is not equal to
'Hola' != 'hola'

True

In [47]:
# minor than
4 < 10

True

### Membership operators

Membership operators are used to test if a sequence is presented in an object such as a string, an integer as well as a tuple.

| Membership Operators | Description  |
| :---: | :---: |
| `in` | if a sequence with the specified value is present in the object, then True |
| `not in` | if a sequence with the specified value is not present in the object, then True  |

Store your surname in a variable and check if surname variable contains specific characters:

In [62]:
# Type the code here:
surname = "Gabbanelli"

'bb' in surname

In [37]:
'in' not in name

True

## Built-in methods for string handling

Store your name in a variable:

In [51]:
# Type the code here:
name = 'Luciano'
name

'Luciano'

`len()` is a built-in function usefull not only for strings, isn't it?

Find the length of your name:

In [38]:
# Type the code here:
len(name)

7

Put your name in all caps:

In [39]:
# Type the code here:
name.upper()

'LUCIANO'

You can ask if all characters in a string are uppercase.

In [55]:
# Type the code here:
name.upper().isupper()

True

Try the following methods: .lower(), .startswith(), .replace(), .capitalize()

Shortcuts :D
Write the name of the variable before a period, then press the tab key. A panel with the available methods will be displayed.

In [25]:
name. # press Tab key next to the dot

SyntaxError: invalid syntax (472924188.py, line 1)

In [57]:
name.lower()

'luciano'

In [58]:
name.startswith('lu')

False

In [59]:
name.removeprefix('Lu')

'ciano'

In [60]:
name.replace('a','e')

'Lucieno'

In [61]:
'anselmo'.capitalize()

'Anselmo'

### Slicing

How strings behave when applying the arithmetic operators + and *

In [2]:
# Type the code here:
a = 'Assembler'
b = 'Institute'

c = a+b
print(c)

AssemblerInstitute


In [45]:
c * 3

'AssemblerInstituteAssemblerInstituteAssemblerInstitute'

In [46]:
c * 3.5

TypeError: can't multiply sequence by non-int of type 'float'

In [3]:
complete_name = a + ' ' + b
complete_name

'Assembler Institute'

You can indicate a range of characters using the split syntax and thus return part of a string.

| A | s | s | e | m | b | l | e | r |  | I | n | s | t | i | t | u | t | e |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 |
| -18 | -17 | -16 | -15 | -14 | -13 | -12 | -11 | -10 | -9 | -8 | -7 | -6 | -5 | -4 | -3 | -2 | -1 | 0 |

**&rarr;** Start to End

&larr; End to Start (for negavive indexing)

In [48]:
complete_name[0]

'A'

Which is the location of the space?

In [49]:
# Type the code here:
complete_name[9]

' '

Python allows **negative indexing** for its sequences (this means start from the end).
The index of -1 refers to the last item, -2 to the second last item and so on. Try it!

In [71]:
complete_name[-1] 

'i'

**Range of Indexes:** You can specify a range of indexes by specifying where to start and where to end the range. The return value will be a new list with the specified items.

Print 'emble', 'Assembler' and ' Institute'

In [50]:
# Type the code here:
complete_name[3:8] # Take care of the endpoints! It would be = [3;8)

'emble'

In [51]:
complete_name[:9]

'Assembler'

In [52]:
complete_name[9:]

' Institute'

And also print 'tute', 'Assembler Inst' and 'Inst' using nevative indexing

In [54]:
# Type the code here:
complete_name[-4:]

'tute'

In [55]:
complete_name[:-5]

'Assembler Inst'

In [57]:
complete_name[-9:-5]

'Inst'

What do you expect from this?

In [60]:
complete_name[:]

'Assembler Institute'

When slicing, 3 parameters are available: $\quad$ `string[start:end:jump]`

So what do you expect the output of the following command to be?

In [61]:
complete_name[2:30:2]

'smlrIsiue'

### The `strip` method

The `strip()` method removes any leading (spaces at the beginning) and trailing (spaces at the end) characters (space is the default leading character to remove)

> **Sitntax**: `string.strip(characters)`
>
> **Parameter**: Optional. Characters to be removed as leading/trailing characters

Two similar methods are `rstrip()` and `lstrip()`, whcich removes any trailing characters (characters at the end a string), or any leading characters, respectively. Space is the default trailing character to remove.

**Task:** Remove leftover characters from the string called `txt` and print the result: 

In [11]:
txt = ",,,,,rrttgg.....banana,,,,,ssqqqww....."

# Type code here:
x = txt.strip(",.qswrtg")
print(x)

banana


### The `split()` method

There is no resume. Go to:

[Python String `split()` Method](https://www.w3schools.com/python/ref_string_split.asp)

**Task:** Split the variable `complete_name` with no arguments.

In [5]:
# Press Shift + Tab inside parenthesis for help
# Type code here:
complete_name.split() 

['Assembler', 'Institute']

**Task:** Split the variable `complete_name` using the character `'t'` as a separator and store the output in a variable. Show the variable, find its type and length.

In [7]:
# Type code here:
splited_complete_name = complete_name.split('t')
splited_complete_name

['Assembler Ins', 'i', 'u', 'e']

In [9]:
type(splited_complete_name)

list

In [10]:
len(splited_complete_name)

4

**Task:** Given the following web page: `'www.assemblerinstitute.com'`, extract the domain (this is, `'assemblerinstitute.com'`)

In [9]:
# Type the code here:
webpage = 'www.assemblerinstitute.com'
webpage.split('.', 1)[1]

### The `.join()` method

Now is your turn! Teach yourself the `.join()` method.

[Python String `join()` Method](https://www.programiz.com/python-programming/methods/string/join)

**Task:** You hace three list with part of a phrase, construct the full quote only using the `join()` method. The output should be:

"What we know is a drop, what we don't know is an ocean. Isaac Newton"

In [ ]:
quote_part1 = ['What', 'we', 'know', 'is', 'a', 'drop']
quote_part2 = ["what", "we", "don't", "know", "is", "an", "ocean"]
author = ['Isaac', 'Newton']

In [17]:
# Type the code here:
string_part1 = ' '.join(quote_part1)
string_part2 = ' '.join(quote_part2)
string_author = ' '.join(author)

In [18]:
string_author

'Isaac Newton'

In [19]:
string_final = ", ".join([string_part1,string_part2])
string_final = ". ".join([string_final,string_author])
string_final

"What we know is a drop, what we don't know is an ocean. Isaac Newton"